src/protocols/bdx/BDXMessageUtils.cpp

```diff
@@ -0,0 +1,444 @@
+    VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    // Metadata is optional
+    if (mMetadata && mMetadataLength > 0)
+    {
+        bbuf.Put(mMetadata, (size_t) mMetadataLength);
+        VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+    }
+
+    aBuffer->SetDataLength(static_cast<uint16_t>(bbuf.Written()));
+
+exit:
+    return err;
+}
+
+CHIP_ERROR TransferInit::Parse(System::PacketBuffer * aBuffer, TransferInit & aParsedMessage)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    uint8_t ptcByte;
+    uint8_t rangeCtl;
+    bool hasStartOffset     = false;
+    uint32_t tmpUint32Value = 0;
+    uint8_t * bufStart;
+
+    // BufBound has to be initialized with aBuffer so VerifyOrExit can't be used here.
+    if (aBuffer == NULL)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    bufStart = aBuffer->Start();
+    Reader bufReader(bufStart, aBuffer->DataLength());
+
+    SuccessOrExit(bufReader.Read8(&ptcByte).StatusCode());
+
+    aParsedMessage.mSupportedVersions     = ptcByte & VERSION_MASK;
+    aParsedMessage.mSupportsSenderDrive   = ((ptcByte & SENDER_DRIVE_MASK) != 0);
+    aParsedMessage.mSupportsReceiverDrive = ((ptcByte & RECEIVER_DRIVE_MASK) != 0);
+    aParsedMessage.mSupportsAsync         = ((ptcByte & ASYNC_MASK) != 0);
+
+    SuccessOrExit(bufReader.Read8(&rangeCtl).StatusCode());
+
+    aParsedMessage.mDefLen    = (rangeCtl & DEFLEN_MASK) != 0;
+    hasStartOffset            = (rangeCtl & START_OFFSET_MASK) != 0;
+    aParsedMessage.mWideRange = (rangeCtl & WIDERANGE_MASK) != 0;
+
+    SuccessOrExit(bufReader.Read16(&aParsedMessage.mMaxBlockSize).StatusCode());
+
+    if (hasStartOffset)
+    {
+        if (aParsedMessage.mWideRange)
+        {
+            SuccessOrExit(bufReader.Read64(&aParsedMessage.mStartOffset).StatusCode());
+        }
+        else
+        {
+            SuccessOrExit(bufReader.Read32(&tmpUint32Value).StatusCode());
+            aParsedMessage.mStartOffset = tmpUint32Value;
+        }
+    }
+
+    if (aParsedMessage.mDefLen)
+    {
+        if (aParsedMessage.mWideRange)
+        {
+            SuccessOrExit(bufReader.Read64(&aParsedMessage.mMaxLength).StatusCode());
+        }
+        else
+        {
+            SuccessOrExit(bufReader.Read32(&tmpUint32Value).StatusCode());
+            aParsedMessage.mMaxLength = tmpUint32Value;
+        }
+    }
+
+    SuccessOrExit(bufReader.Read16(&aParsedMessage.mFileDesLength).StatusCode());
+
+    VerifyOrExit(bufReader.OctetsRead() + aParsedMessage.mFileDesLength <= aBuffer->DataLength(),
+                 err = CHIP_ERROR_MESSAGE_INCOMPLETE);
+    aParsedMessage.mFileDesignator = &bufStart[bufReader.OctetsRead()];
+
+    // Rest of message is metadata (could be empty)
+    if (bufReader.OctetsRead() + aParsedMessage.mFileDesLength < aBuffer->DataLength())
+    {
+        uint16_t metadataStartIndex    = (uint16_t)(bufReader.OctetsRead() + aParsedMessage.mFileDesLength);
```

Comment:
C++-style casts, please, here and elsewhere.

Reply:
Will do.